### PR TITLE
 *: add split table syntax to split table region

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Please merge :https://github.com/pingcap/tidb/pull/10409 first.
Add new syntax for `split table region`.

```sql
SPLIT TABLE [table_name] BETWEEN (min_value...) AND (max_value...) REGIONS [region_num]

SPLIT TABLE [table_name] BY [value_lists]
```

#### Attention

- The `min` and `max` value count should be 1.
- The max split index region num currently is `1000` in a single query.

**Eg1**

```sql
split table `t` min (0) max (1000) num 5;
```

Upper sql will split `5` regions of the table `t`  in range `0~1000`. Every region range like below:

Region1: -inf ~ 200

Region2: 200 ~ 400

Region3: 400 ~ 600

Region4: 600 ~ 800

Region5: 800 ~ +inf

**Eg2**

```sql
split table `t` by (100),(200),(300);
```
Upper sql will split `4` regions of the table `t`  , Every region range like below:

Region1: -inf ~ 100

Region2: 100 ~ 200

Region3: 200 ~ 300

Region4: 300 ~ +inf

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression

Related changes

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -1255,13 +1255,26 @@ func (b *executorBuilder) buildUnionAll(v *plannercore.PhysicalUnionAll) Executo
 	return e
 }
 
-func (b *executorBuilder) buildSplitIndexRegion(v *plannercore.SplitIndexRegion) Executor {
+func (b *executorBuilder) buildSplitIndexRegion(v *plannercore.SplitRegion) Executor {
```
Comment (winkyao) [inline executor/builder.go]: buildSplitIndexRegion --> buildSplitRegion

File: executor/executor_test.go [hunk 2]
```diff
@@ -3831,7 +3832,44 @@ func (s *testSuite) TestSplitIndexRegion(c *C) {
 	// Test truncate error msg.
 	_, err = tk.Exec(`split table t index idx1 between ("aa") and (1000000000) regions 0`)
 	c.Assert(err, NotNil)
-	c.Assert(err.Error(), Equals, "[types:1265]Incorrect value: 'aa' for index column 'b'")
+	c.Assert(err.Error(), Equals, "[types:1265]Incorrect value: 'aa' for column 'b'")
+
+	// Test for split table region.
+	tk.MustExec(`split table t between (0) and (1000000000) regions 10`)
+	// Check min value is more than upper value.
```
Comment (bb7133) [inline executor/executor_test.go]:
```suggestion
	// Check the ower value is more than the upper value.
```

File: executor/executor_test.go [hunk 3]
```diff
@@ -3831,7 +3832,44 @@ func (s *testSuite) TestSplitIndexRegion(c *C) {
...
+	_, err = tk.Exec(`split table t between (2) and (1) regions 10`)
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "Split table t region lower value 2 should less than the upper value 1")
+
+	// Check min value is invalid.
```
Comment (bb7133) [inline executor/executor_test.go]:
```suggestion
	// Check the lower value is invalid.
```

File: executor/split.go [hunk 4]
```diff
@@ -204,3 +205,107 @@ func datumSliceToString(ds []types.Datum) (string, error) {
...
+	s, ok := store.(splitableStore)
+	if !ok {
+		return nil
+	}
+	splitKeys, err := e.getSplitTableKeys()
+	if err != nil {
+		return err
+	}
+	regionIDs := make([]uint64, 0, len(splitKeys))
+	for _, idxKey := range splitKeys {
```
Comment (bb7133) [inline executor/split.go]:
```suggestion
	for _, key := range splitKeys {
```

File: planner/core/planbuilder.go [hunk 5]
```diff
@@ -1686,43 +1693,110 @@ func (b *PlanBuilder) buildSplitRegion(node *ast.SplitRegionStmt) (Plan, error)
...
+		expr, _, err = b.rewrite(valueItem, mockTablePlan, nil, true)
 		if err != nil {
-			if !types.ErrTruncated.Equal(err) {
-				return nil, err
+			return d, err
+		}
+	}
+	constant, ok := expr.(*expression.Constant)
+	if !ok {
+		return d, errors.New("expect constant values")
```
Comment (bb7133) [inline planner/core/planbuilder.go]:
```suggestion
		return d, errors.New("Expect constant values.")
```

File: executor/split.go (hunk 4 above)
Comment (crazycs520) [reply executor/split.go]: done.

**Comment by bb7133** [comment]
However, we should take care of the English syntax in self-made error messages. Different from code variables/comments, error messages are sent and shown to end-users.

I advise making another PR to fix typos/misspellings in this PR and https://github.com/pingcap/tidb/pull/10409, for example:

https://github.com/pingcap/tidb/pull/10553/files#diff-ce300a40f7a5a1103912bc56991b9cb8R1794

File: executor/split.go [hunk 6]
```diff
@@ -204,3 +205,107 @@ func datumSliceToString(ds []types.Datum) (string, error) {
...
+		return keys, nil
+	}
+	isUnsigned := false
+	if e.tableInfo.PKIsHandle {
+		if pkCol := e.tableInfo.GetPkColInfo(); pkCol != nil {
+			isUnsigned = mysql.HasUnsignedFlag(pkCol.Flag)
+		}
+	}
+	step := uint64(0)
+	lowerValue := int64(0)
```
Comment (winkyao) [inline executor/split.go]: var lowerValue int64

File: executor/split.go [hunk 7]
```diff
@@ -204,3 +205,107 @@ func datumSliceToString(ds []types.Datum) (string, error) {
 	str += ")"
 	return str, nil
 }
+
+// SplitTableRegionExec represents a split table regions executor.
+type SplitTableRegionExec struct {
+	baseExecutor
+
+	tableInfo  *model.TableInfo
+	lower      []types.Datum
```
Comment (winkyao) [inline executor/split.go]: lower and upper can be `types.Datum` not `[]types.Datum`

File: executor/split.go [hunk 8]
```diff
@@ -204,3 +205,107 @@ func datumSliceToString(ds []types.Datum) (string, error) {
...
+	for _, key := range splitKeys {
+		regionID, err := s.SplitRegionAndScatter(key)
+		if err != nil {
+			logutil.Logger(context.Background()).Warn("split table region failed",
+				zap.String("table", e.tableInfo.Name.L),
+				zap.Error(err))
+			continue
+		}
+		regionIDs = append(regionIDs, regionID)
+
```
Comment (zimulala) [inline executor/split.go]: Remove this useless line.

**Pull request merged**